source/aggregation.txt

```diff
@@ -0,0 +1,190 @@
+domain-specific language (DSL). The DSL supports the following aggregation
+pipeline operators:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 50 50
+    
+   * - Operator
+     - Method Name
+   
+   * - :manual:`$group <reference/operator/aggregation/group/>`
+     - ``group()``
+
+   * - :manual:`$project <reference/operator/aggregation/project/>`
+     - ``project()``
+
+   * - :manual:`$unwind <reference/operator/aggregation/unwind/>`
+     - ``unwind()``
+
+To create an aggregation pipeline by using one of the preceding operators, call
+the corresponding method on an instance of ``Criteria``. Calling the method adds
+the aggregation operation to the ``pipeline`` atrritbure of the ``Criteria``
+instance. To run the aggregation pipeline, pass the ``pipeline`` attribute value
+to the ``Collection#aggregate()`` method. 
+
+Example
+~~~~~~~
+
+Consider a collection that contains documents that are modeled as follows:
+
+.. code-block:: ruby
+
+   class Tour
+     include Mongoid::Document
+
+     embeds_many :participants
+
+     field :name, type: String
+     field :states, type: Array
+   end
+
+   class Participant
+     include Mongoid::Document
+
+     embedded_in :tour
+
+     field :name, type: String
+   end
+
+The following example creates an aggregation pipeline that outputs the states a
+participant has visited by using the following
+aggregation operations:
+
+- ``$match``, which find documents in which the ``participants.name`` field
+  value is ``"Serenity"``
+- ``$unwind``, which deconstructs the ``states`` array field and outputs a
+  document for each element in the array
+- ``$group``, which groups the documents by the value of their ``states`` field
+- ``$project``, which prompts the pipeline to return only the ``_id`` and
+  ``states`` fields
+
+.. literalinclude:: /includes/aggregation/builder-dsl.rb
+   :language: ruby
+
+Aggregation with the Ruby Driver
+--------------------------------
+
+You can use the Ruby driver to run aggregation operations that do not have
+corresponding builder methods by using the
+``Collection#aggregate()`` method and passing in an array of aggregation
+operations. Creating the aggregation pipeline by using the Ruby driver returns
+raw ``BSON::Document`` objects rather than ``Mongoid::Document`` model
+instances.
+
+Example
+~~~~~~~
+
+Consider a collection that contains documents that are modeled as follows:
+
+.. code-block:: ruby
+    
+   class Band
+     include Mongoid::Document
+     has_many :tours
+     has_many :awards
+     field :name, type: String
+   end
+
+   class Tour
+     include Mongoid::Document
+     belongs_to :band
+     field :year, type: Integer
+   end
+
+   class Award
+     include Mongoid::Document
+     belongs_to :band
+     field :name, type: String
+   end
+
+The following example creates an aggregation pipeline to retrieve all bands that
+have toured since 2000 and have at least one award:
```

Comment:
```suggestion
The following example creates an aggregation pipeline to retrieve all bands that
have toured since ``2000`` and have at least ``1`` award:
```